### PR TITLE
[RLlib] Cleanup some deprecated metric keys and classes.

### PR DESCRIPTION
**Pull request by sven1977**
This PR:
* deprecates TimerStats (in favor of ray.util.timer::_Timer).
* reduces usage of already deprecated metrics keys, such as: `WORKER_UPDATE_TIMER`, `STEPS_TRAINED_COUNTER`, `STEPS_TRAINED_THIS_ITER_COUNTER`, etc.. These have already been (soft) deprecated a while back, but are still inconsistently being used in some `training_step()` methods, leading to faulty results dicts.
* Bug fix: Error in counting evaluation (env and agent) steps per training iteration if `evaluation_duration=auto`.
* Bug fix: RLlib would not necessarily respect the "episode smoothing window" setting, in certain scenarios.
* Get rid of `Algorithm._rollout_worker_metrics` buffer (not required any longer).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/a2c/a2c.py [hunk 1]
```diff
@@ -188,8 +186,8 @@ def training_step(self) -> ResultDict:
         )
         if self._num_microbatches >= num_microbatches:
             # Update counters.
-            self._counters[STEPS_TRAINED_COUNTER] += self._microbatches_counts
-            self._counters[STEPS_TRAINED_THIS_ITER_COUNTER] = self._microbatches_counts
+            self._counters[NUM_ENV_STEPS_TRAINED] += self._microbatches_counts
```
Comment (sven1977) [inline rllib/algorithms/a2c/a2c.py:189]: Stop using deprecated names.

File: rllib/algorithms/algorithm.py [hunk 2]
```diff
@@ -593,13 +595,12 @@ def step(self) -> ResultDict:
             results.update(self._run_one_evaluation(train_future=None))
 
         # Collect rollout worker metrics.
-        episodes, self._episodes_to_be_collected = collect_episodes(
+        episodes_this_iter, self._episodes_to_be_collected = collect_episodes(
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:598]: clarify var names

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -780,19 +782,22 @@ def duration_fn(num_units_done):
                             < units_left_to_do
                         ]
                     )
-                    agent_steps_this_iter = sum(b.agent_steps() for b in batches)
-                    env_steps_this_iter = sum(b.env_steps() for b in batches)
+                    _agent_steps = sum(b.agent_steps() for b in batches)
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:785]: this was a bug, leading to wrong (fewer) counts

File: rllib/algorithms/algorithm.py [hunk 4]
```diff
@@ -2382,18 +2389,29 @@ def _compile_iteration_results(self, *, step_ctx, iteration_results=None):
         # Learner info.
         results["info"] = {LEARNER_INFO: iteration_results}
 
-        episodes = self._rollout_worker_metrics
-        orig_episodes = list(episodes)
-        missing = self.config["metrics_num_episodes_for_smoothing"] - len(episodes)
+        # Calculate how many (if any) of older, historical episodes we have to add to
+        # `episodes_this_iter` in order to reach the required smoothing window.
+        episodes_for_metrics = episodes_this_iter[:]
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2394]: Simplify code: `self._rollout_worker_metrics` is no longer needed.

File: rllib/algorithms/algorithm.py [hunk 5]
```diff
@@ -2413,11 +2431,15 @@ def _compile_iteration_results(self, *, step_ctx, iteration_results=None):
             results[NUM_AGENT_STEPS_TRAINED + "_this_iter"] = step_ctx.trained
             # TODO: For CQL and other algos, count by trained steps.
             results["timesteps_total"] = self._counters[NUM_AGENT_STEPS_SAMPLED]
+            # TODO: Backward compatibility.
+            results[STEPS_TRAINED_THIS_ITER_COUNTER] = step_ctx.trained
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2435]: Make sure old keys are still available in the metrics, but do NOT require user to keep track of these in the `training_step()` methods.

File: rllib/algorithms/apex_dqn/apex_dqn.py [hunk 6]
```diff
@@ -517,6 +514,9 @@ def update_workers(self, _num_samples_ready: Dict[ActorHandle, int]) -> int:
             self.learner_thread.weights_updated = False
             weights = self.workers.local_worker().get_weights()
             self.curr_learner_weights = ray.put(weights)
+
+        num_workers_updated = 0
```
Comment (sven1977) [inline rllib/algorithms/apex_dqn/apex_dqn.py:518]: fix: method is supposed to return number of workers updated, but didn't.

File: rllib/algorithms/apex_dqn/apex_dqn.py [hunk 7]
```diff
@@ -617,7 +621,6 @@ def update_replay_sample_priority(self) -> None:
             else:
                 raise RuntimeError("The learner thread died in while training")
 
-        self._counters[STEPS_TRAINED_THIS_ITER_COUNTER] = num_samples_trained_this_itr
```
Comment (sven1977) [inline rllib/algorithms/apex_dqn/apex_dqn.py:620]: Deprecated metrics key. Will be handled automatically by Algorithm.

File: rllib/algorithms/ddppo/ddppo.py [hunk 8]
```diff
@@ -24,9 +24,6 @@
 from ray.rllib.algorithms.ppo import PPOConfig, PPO
 from ray.rllib.evaluation.postprocessing import Postprocessing
 from ray.rllib.evaluation.rollout_worker import RolloutWorker
-from ray.rllib.execution.common import (
```
Comment (sven1977) [inline rllib/algorithms/ddppo/ddppo.py:27]: automated:
deprecated metrics keys should no longer be handled by users in `training_step()` methods

File: rllib/algorithms/dreamer/dreamer.py [hunk 9]
```diff
@@ -383,10 +387,11 @@ def training_step(self) -> ResultDict:
         # Number of sub-iterations for Dreamer
         dreamer_train_iters = self.config["dreamer_train_iters"]
         batch_size = self.config["batch_size"]
-        action_repeat = self.config["action_repeat"]
 
         # Collect SampleBatches from rollout workers.
         batch = synchronous_parallel_sample(worker_set=self.workers)
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += batch.agent_steps()
```
Comment (sven1977) [inline rllib/algorithms/dreamer/dreamer.py:393]: this bookkeeping step was simply missing

File: rllib/algorithms/impala/impala.py [hunk 10]
```diff
@@ -874,17 +874,17 @@ def process_experiences_tree_aggregation(
     def update_workers_if_necessary(self) -> None:
         # Only need to update workers if there are remote workers.
         global_vars = {"timestep": self._counters[NUM_AGENT_STEPS_TRAINED]}
-        self._counters["steps_since_broadcast"] += 1
+        self._counters["num_training_step_calls_since_last_broadcast"] += 1
```
Comment (sven1977) [inline rllib/algorithms/impala/impala.py:877]: Clarify name. "steps" is very confusing (env steps? agent steps?). We mean to count the number of times we have called `training_step()` here.

File: rllib/algorithms/impala/impala.py [hunk 11]
```diff
@@ -874,17 +874,17 @@ def process_experiences_tree_aggregation(
...
+            and self._counters["num_training_step_calls_since_last_broadcast"]
             >= self.config["broadcast_interval"]
             and self.workers_that_need_updates
         ):
             weights = ray.put(self.workers.local_worker().get_weights())
-            self._counters["steps_since_broadcast"] = 0
+            self._counters["num_training_step_calls_since_last_broadcast"] = 0
             self._learner_thread.weights_updated = False
-            self._counters["num_weight_broadcasts"] += 1
+            self._counters[NUM_SYNCH_WORKER_WEIGHTS] += 1
```
Comment (sven1977) [inline rllib/algorithms/impala/impala.py:887]: Use unified name

File: rllib/algorithms/algorithm.py [hunk 12]
```diff
@@ -2382,18 +2389,29 @@ def _compile_iteration_results(self, *, step_ctx, iteration_results=None):
...
+            )
+        # Add new episodes to our history and make sure it doesn't grow larger than
+        # needed.
+        self._episode_history.extend(episodes_this_iter)
         self._episode_history = self._episode_history[
             -self.config["metrics_num_episodes_for_smoothing"] :
         ]
         results["sampler_results"] = summarize_episodes(
-            episodes, orig_episodes, self.config["keep_per_episode_custom_metrics"]
+            episodes_for_metrics,
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2412]: Clarified var names.

File: rllib/algorithms/algorithm.py [hunk 13]
```diff
@@ -613,9 +614,10 @@ def step(self) -> ResultDict:
             # Sync filters on workers.
             self._sync_filters_if_needed(self.workers)
 
-            # Collect worker metrics.
+            # Collect worker metrics and add combine them with `results`.
```
Comment (gjoliver) [inline rllib/algorithms/algorithm.py:617]: nit, typo, '... and combine them ...'?

File: rllib/algorithms/apex_dqn/apex_dqn.py [hunk 14]
```diff
@@ -657,10 +668,8 @@ def on_worker_failures(
             self._sampling_actor_manager.add_workers(new_workers)
 
     @override(Algorithm)
-    def _compile_iteration_results(self, *, step_ctx, iteration_results=None):
-        result = super()._compile_iteration_results(
-            step_ctx=step_ctx, iteration_results=iteration_results
-        )
+    def _compile_iteration_results(self, *args, **kwargs):
```
Comment (gjoliver) [inline rllib/algorithms/apex_dqn/apex_dqn.py:671]: curious, why change func arguments to this generic types?

File: rllib/algorithms/impala/impala.py (hunk 10 above)
Comment (gjoliver) [reply rllib/algorithms/impala/impala.py:877]: this is a good idea. can we create a name for this string in rllib.utils.metrics?

**Review by gjoliver — COMMENTED** [review]
cool!! just a few really minor questions and comments.

File: rllib/algorithms/algorithm.py (hunk 13 above)
Comment (sven1977) [inline rllib/algorithms/algorithm.py:617]:
```suggestion
            # Collect worker metrics and combine them with `results`.
```

File: rllib/algorithms/algorithm.py [hunk 15]
```diff
@@ -780,19 +782,20 @@ def duration_fn(num_units_done):
...
                     else:
                         num_units_done += (
-                            agent_steps_this_iter
-                            if self._by_agent_steps
-                            else env_steps_this_iter
+                            _agent_steps if self._by_agent_steps else _env_steps
                         )
 
+                    agent_steps_this_iter += _agent_steps
+                    env_steps_this_iter += _env_steps
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:797]: Yes, this was a real bug! Affected all algorithms that call `training_step()` multiple times to make up one iteration (== a single call to `.train()`), e.g. the async ones.

File: rllib/algorithms/apex_dqn/apex_dqn.py (hunk 14 above)
Comment (sven1977) [reply rllib/algorithms/apex_dqn/apex_dqn.py:671]: To make things less prone to forward compat errors in the future.
If the args are not really needed in the child implementation, then why not? :)

**Comment by sven1977** [comment]
Hey @gjoliver , thanks for the review. I made the changes you requested.
I think a next step would be to create a `TrainingStepResults` object that includes sub-results objects so we can get rid of these annoying dict keys. Only custom metrics should use dictionaries, IMO.
After that, we should properly document this.
I feel like it's one of the last huge mysteries for users when they need to customize the `training_step()` method. "What do I even need to return from this method?" and "Where does RLlib magically get the rollout metrics from?" :)

File: rllib/algorithms/apex_dqn/apex_dqn.py (hunk 14 above)
Comment (gjoliver) [reply rllib/algorithms/apex_dqn/apex_dqn.py:671]: I see. a minor suggestion may be to name the arguments *unused_args and **unused_kwargs to make the intention clearer.
if in the future we do need some parameters, it's better to write it out.

**Review by gjoliver — APPROVED** [review]
ok, one nit for your consideration, but thanks a ton for the fix!!

**Pull request merged**